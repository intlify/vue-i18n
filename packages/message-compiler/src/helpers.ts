export const enum HelperNameMap {
  LIST = 'list',
  NAMED = 'named',
  PLURAL = 'plural',
  LINKED = 'linked',
  MESSAGE = 'message',
  TYPE = 'type',
  INTERPOLATE = 'interpolate',
  NORMALIZE = 'normalize'
}
