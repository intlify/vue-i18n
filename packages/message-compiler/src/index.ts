export * from './options'
export * from './errors'
export * from './helpers'
export * from './location'
export * from './parser'
export * from './compiler'
