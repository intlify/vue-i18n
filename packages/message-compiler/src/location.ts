// The source range.
// The `start` is inclusive and `end` is exclusive.
// [start, end)
export interface SourceLocation {
  start: Position
  end: Position
  source?: string
}

export const LocationStub: SourceLocation = {
  start: { line: 1, column: 1, offset: 0 },
  end: { line: 1, column: 1, offset: 0 }
}

export interface Position {
  offset: number // from start of content (file)
  line: number
  column: number
}

export function createPosition(
  line: number,
  column: number,
  offset: number
): Position {
  return { line, column, offset } as Position
}

export function createLocation(
  start: Position,
  end: Position,
  source?: string
): SourceLocation {
  const loc = { start, end } as SourceLocation
  if (source != null) {
    loc.source = source
  }
  return loc
}
