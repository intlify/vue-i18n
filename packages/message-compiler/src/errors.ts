import { format } from '@intlify/shared'
import { SourceLocation } from './location'

export type CompileDomain =
  | 'tokenizer'
  | 'parser'
  | 'generator'
  | 'transformer'
  | 'compiler'

export interface CompileError extends SyntaxError {
  code: number
  domain?: CompileDomain
  location?: SourceLocation
}

export interface CreateCompileErrorOptions {
  domain?: CompileDomain
  messages?: { [code: number]: string }
  args?: unknown[]
}

export const enum CompileErrorCodes {
  // tokenizer error codes
  EXPECTED_TOKEN,
  INVALID_TOKEN_IN_PLACEHOLDER,
  UNTERMINATED_SINGLE_QUOTE_IN_PLACEHOLDER,
  UNKNOWN_ESCAPE_SEQUENCE,
  INVALID_UNICODE_ESCAPE_SEQUENCE,
  UNBALANCED_CLOSING_BRACE,
  UNTERMINATED_CLOSING_BRACE,
  EMPTY_PLACEHOLDER,
  NOT_ALLOW_NEST_PLACEHOLDER,
  INVALID_LINKED_FORMAT,

  // parser error codes
  MUST_HAVE_MESSAGES_IN_PLURAL,
  UNEXPECTED_EMPTY_LINKED_MODIFIER,
  UNEXPECTED_EMPTY_LINKED_KEY,
  UNEXPECTED_LEXICAL_ANALYSIS,

  // Special value for higher-order compilers to pick up the last code
  // to avoid collision of error codes. This should always be kept as the last
  // item.
  __EXTEND_POINT__
}

/** @internal */
export const errorMessages: { [code: number]: string } = {
  // tokenizer error messages
  [CompileErrorCodes.EXPECTED_TOKEN]: `Expected token: '{0}'`,
  [CompileErrorCodes.INVALID_TOKEN_IN_PLACEHOLDER]: `Invalid token in placeholder: '{0}'`,
  [CompileErrorCodes.UNTERMINATED_SINGLE_QUOTE_IN_PLACEHOLDER]: `Unterminated single quote in placeholder`,
  [CompileErrorCodes.UNKNOWN_ESCAPE_SEQUENCE]: `Unknown escape sequence: \\{0}`,
  [CompileErrorCodes.INVALID_UNICODE_ESCAPE_SEQUENCE]: `Invalid unicode escape sequence: {0}`,
  [CompileErrorCodes.UNBALANCED_CLOSING_BRACE]: `Unbalanced closing brace`,
  [CompileErrorCodes.UNTERMINATED_CLOSING_BRACE]: `Unterminated closing brace`,
  [CompileErrorCodes.EMPTY_PLACEHOLDER]: `Empty placeholder`,
  [CompileErrorCodes.NOT_ALLOW_NEST_PLACEHOLDER]: `Not allowed nest placeholder`,
  [CompileErrorCodes.INVALID_LINKED_FORMAT]: `Invalid linked format`,
  // parser error messages
  [CompileErrorCodes.MUST_HAVE_MESSAGES_IN_PLURAL]: `Plural must have messages`,
  [CompileErrorCodes.UNEXPECTED_EMPTY_LINKED_MODIFIER]: `Unexpected empty linked modifier`,
  [CompileErrorCodes.UNEXPECTED_EMPTY_LINKED_KEY]: `Unexpected empty linked key`,
  [CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS]: `Unexpected lexical analysis in token: '{0}'`
}

export function createCompileError<T extends number>(
  code: T,
  loc: SourceLocation | null,
  options: CreateCompileErrorOptions = {}
): CompileError {
  const { domain, messages, args } = options
  const msg = __DEV__
    ? format((messages || errorMessages)[code] || '', ...(args || []))
    : code
  const error = new SyntaxError(String(msg)) as CompileError
  error.code = code
  if (loc) {
    error.location = loc
  }
  error.domain = domain
  return error
}

/** @internal */
export function defaultOnError(error: CompileError): never {
  throw error
}
