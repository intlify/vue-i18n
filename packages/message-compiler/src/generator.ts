import { isString } from '@intlify/shared'
import { SourceMapGenerator, RawSourceMap } from 'source-map'
import {
  ResourceNode,
  Node,
  NodeTypes,
  PluralNode,
  MessageNode,
  TextNode,
  ListNode,
  NamedNode,
  LinkedNode,
  LinkedKeyNode,
  LinkedModitierNode,
  LiteralNode
} from './parser'
import { Position, LocationStub } from './location'
import { CodeGenOptions } from './options'
import { HelperNameMap } from './helpers'

export interface CodeGenResult {
  code: string
  ast: ResourceNode
  map?: RawSourceMap
}

type CodeGenContext = {
  source: string
  code: string
  indentLevel: number
  filename: string
  line: number
  breakLineCode: string
  needIndent: boolean
  column: number
  offset: number
  map?: SourceMapGenerator
}

type CodeGenNode =
  | TextNode
  | NamedNode
  | ListNode
  | LiteralNode
  | LinkedKeyNode
  | LinkedModitierNode

type CodeGenerator = {
  context(): CodeGenContext
  push(code: string, node?: CodeGenNode): void
  indent(withNewLine?: boolean): void
  deindent(withNewLine?: boolean): void
  newline(): void
  helper(key: string): string
  needIndent(): boolean
}

function createCodeGenerator(
  ast: ResourceNode,
  options: CodeGenOptions
): CodeGenerator {
  const {
    sourceMap,
    filename,
    breakLineCode,
    needIndent: _needIndent
  } = options
  const _context = {
    source: ast.loc!.source,
    filename,
    code: '',
    column: 1,
    line: 1,
    offset: 0,
    map: undefined,
    breakLineCode,
    needIndent: _needIndent,
    indentLevel: 0
  } as CodeGenContext

  const context = (): CodeGenContext => _context

  function push(code: string, node?: CodeGenNode): void {
    _context.code += code
    if (!__BROWSER__ && _context.map) {
      if (node && node.loc && node.loc !== LocationStub) {
        addMapping(node.loc.start, getMappingName(node))
      }
      advancePositionWithSource(_context, code)
    }
  }

  function _newline(n: number, withBreakLine = true): void {
    const _breakLineCode = withBreakLine ? breakLineCode : ''
    push(_needIndent ? _breakLineCode! + `  `.repeat(n) : _breakLineCode!)
  }

  function indent(withNewLine = true): void {
    const level = ++_context.indentLevel
    withNewLine && _newline(level)
  }

  function deindent(withNewLine = true): void {
    const level = --_context.indentLevel
    withNewLine && _newline(level)
  }

  function newline(): void {
    _newline(_context.indentLevel)
  }

  const helper = (key: string): string => `_${key}`
  const needIndent = (): boolean => _context.needIndent

  function addMapping(loc: Position, name?: string) {
    _context.map!.addMapping({
      name,
      source: _context.filename,
      original: {
        line: loc.line,
        column: loc.column - 1
      },
      generated: {
        line: _context.line,
        column: _context.column - 1
      }
    })
  }

  if (!__BROWSER__ && sourceMap) {
    _context.map = new SourceMapGenerator()
    _context.map!.setSourceContent(filename!, _context.source)
  }

  return {
    context,
    push,
    indent,
    deindent,
    newline,
    helper,
    needIndent
  }
}

function generateLinkedNode(generator: CodeGenerator, node: LinkedNode): void {
  const { helper } = generator
  generator.push(`${helper(HelperNameMap.LINKED)}(`)
  generateNode(generator, node.key)
  if (node.modifier) {
    generator.push(`, `)
    generateNode(generator, node.modifier)
  }
  generator.push(`)`)
}

function generateMessageNode(
  generator: CodeGenerator,
  node: MessageNode
): void {
  const { helper, needIndent } = generator
  generator.push(`${helper(HelperNameMap.NORMALIZE)}([`)
  generator.indent(needIndent())
  const length = node.items.length
  for (let i = 0; i < length; i++) {
    generateNode(generator, node.items[i])
    if (i === length - 1) {
      break
    }
    generator.push(', ')
  }
  generator.deindent(needIndent())
  generator.push('])')
}

function generatePluralNode(generator: CodeGenerator, node: PluralNode): void {
  const { helper, needIndent } = generator
  if (node.cases.length > 1) {
    generator.push(`${helper(HelperNameMap.PLURAL)}([`)
    generator.indent(needIndent())
    const length = node.cases.length
    for (let i = 0; i < length; i++) {
      generateNode(generator, node.cases[i])
      if (i === length - 1) {
        break
      }
      generator.push(', ')
    }
    generator.deindent(needIndent())
    generator.push(`])`)
  }
}

function generateResource(generator: CodeGenerator, node: ResourceNode): void {
  if (node.body) {
    generateNode(generator, node.body)
  } else {
    generator.push('null')
  }
}

function generateNode(generator: CodeGenerator, node: Node): void {
  const { helper } = generator
  switch (node.type) {
    case NodeTypes.Resource:
      generateResource(generator, node as ResourceNode)
      break
    case NodeTypes.Plural:
      generatePluralNode(generator, node as PluralNode)
      break
    case NodeTypes.Message:
      generateMessageNode(generator, node as MessageNode)
      break
    case NodeTypes.Linked:
      generateLinkedNode(generator, node as LinkedNode)
      break
    case NodeTypes.LinkedModifier:
      generator.push(
        JSON.stringify((node as LinkedModitierNode).value),
        node as LinkedModitierNode
      )
      break
    case NodeTypes.LinkedKey:
      generator.push(
        JSON.stringify((node as LinkedKeyNode).value),
        node as LinkedKeyNode
      )
      break
    case NodeTypes.List:
      generator.push(
        `${helper(HelperNameMap.INTERPOLATE)}(${helper(HelperNameMap.LIST)}(${
          (node as ListNode).index
        }))`,
        node as ListNode
      )
      break
    case NodeTypes.Named:
      generator.push(
        `${helper(HelperNameMap.INTERPOLATE)}(${helper(
          HelperNameMap.NAMED
        )}(${JSON.stringify((node as NamedNode).key)}))`,
        node as NamedNode
      )
      break
    case NodeTypes.Literal:
      generator.push(
        JSON.stringify((node as LiteralNode).value),
        node as LiteralNode
      )
      break
    case NodeTypes.Text:
      generator.push(JSON.stringify((node as TextNode).value), node as TextNode)
      break
    default:
      if (__DEV__) {
        throw new Error(`unhandled codegen node type: ${node.type}`)
      }
  }
}

// generate code from AST
export const generate = (
  ast: ResourceNode,
  options: CodeGenOptions = {} // eslint-disable-line
): CodeGenResult => {
  const mode = isString(options.mode) ? options.mode : 'normal'
  const filename = isString(options.filename)
    ? options.filename
    : 'message.intl'
  const sourceMap = !!options.sourceMap
  // prettier-ignore
  const breakLineCode = options.breakLineCode != null
    ? options.breakLineCode
    : mode === 'arrow'
      ? ';'
      : '\n'
  const needIndent = options.needIndent ? options.needIndent : mode !== 'arrow'
  const helpers = ast.helpers || []
  const generator = createCodeGenerator(ast, {
    mode,
    filename,
    sourceMap,
    breakLineCode,
    needIndent
  })

  generator.push(mode === 'normal' ? `function __msg__ (ctx) {` : `(ctx) => {`)
  generator.indent(needIndent)

  if (helpers.length > 0) {
    generator.push(
      `const { ${helpers.map(s => `${s}: _${s}`).join(', ')} } = ctx`
    )
    generator.newline()
  }

  generator.push(`return `)
  generateNode(generator, ast)
  generator.deindent(needIndent)
  generator.push(`}`)

  const { code, map } = generator.context()
  return {
    ast,
    code,
    map: map ? (map as any).toJSON() : undefined // eslint-disable-line @typescript-eslint/no-explicit-any
  }
}

function getMappingName(node: CodeGenNode): string | undefined {
  switch (node.type) {
    case NodeTypes.Text:
      return node.value
    case NodeTypes.List:
      return node.index.toString()
    case NodeTypes.Named:
      return node.key
    case NodeTypes.Literal:
      return node.value
    case NodeTypes.LinkedModifier:
      return node.value
    case NodeTypes.LinkedKey:
      return node.value
    default:
      return undefined
  }
}

function advancePositionWithSource(
  pos: Position,
  source: string,
  numberOfCharacters: number = source.length
): Position {
  let linesCount = 0
  let lastNewLinePos = -1
  for (let i = 0; i < numberOfCharacters; i++) {
    if (source.charCodeAt(i) === 10 /* newline char code */) {
      linesCount++
      lastNewLinePos = i
    }
  }

  pos.offset += numberOfCharacters
  pos.line += linesCount
  pos.column =
    lastNewLinePos === -1
      ? pos.column + numberOfCharacters
      : numberOfCharacters - lastNewLinePos

  return pos
}
