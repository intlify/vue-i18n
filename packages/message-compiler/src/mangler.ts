import { CompileErrorCodes, createCompileError } from './errors'
import { NodeTypes } from './nodes'

import type {
  LinkedKeyNode,
  LinkedModifierNode,
  LinkedNode,
  ListNode,
  LiteralNode,
  MessageNode,
  NamedNode,
  Node,
  PluralNode,
  ResourceNode,
  TextNode
} from './nodes'

export const ERROR_DOMAIN = 'minifier'

/* eslint-disable @typescript-eslint/no-explicit-any */

export function mangle(node: Node) {
  node.t = node.type

  switch (node.type) {
    case NodeTypes.Resource: {
      const resource = node as ResourceNode
      mangle(resource.body)
      resource.b = resource.body
      delete (resource as any).body
      break
    }
    case NodeTypes.Plural: {
      const plural = node as PluralNode
      const cases = plural.cases
      for (let i = 0; i < cases.length; i++) {
        mangle(cases[i])
      }
      plural.c = cases
      delete (plural as any).cases
      break
    }
    case NodeTypes.Message: {
      const message = node as MessageNode
      const items = message.items
      for (let i = 0; i < items.length; i++) {
        mangle(items[i])
      }
      message.i = items
      delete (message as any).items
      if (message.static) {
        message.s = message.static
        delete (message as any).static
      }
      break
    }
    case NodeTypes.Text:
    case NodeTypes.Literal:
    case NodeTypes.LinkedModifier:
    case NodeTypes.LinkedKey: {
      const valueNode = node as
        | TextNode
        | LiteralNode
        | LinkedKeyNode
        | LinkedModifierNode
      if (valueNode.value) {
        valueNode.v = valueNode.value
        delete (valueNode as any).value
      }
      break
    }
    case NodeTypes.Linked: {
      const linked = node as LinkedNode
      mangle(linked.key)
      linked.k = linked.key
      delete (linked as any).key
      if (linked.modifier) {
        mangle(linked.modifier)
        linked.m = linked.modifier
        delete (linked as any).modifier
      }
      break
    }
    case NodeTypes.List: {
      const list = node as ListNode
      list.i = list.index
      delete (list as any).index
      break
    }
    case NodeTypes.Named: {
      const named = node as NamedNode
      named.k = named.key
      delete (named as any).key
      break
    }
    default:
      if (__DEV__) {
        throw createCompileError(
          CompileErrorCodes.UNHANDLED_MINIFIER_NODE_TYPE,
          null,
          {
            domain: ERROR_DOMAIN,
            args: [node.type]
          }
        )
      }
  }

  delete (node as any).type
}

/* eslint-enable @typescript-eslint/no-explicit-any */
