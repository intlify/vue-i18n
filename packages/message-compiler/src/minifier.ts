import { NodeTypes } from './nodes'
import { createCompileError, CompileErrorCodes } from './errors'

import type {
  MessageNode,
  ResourceNode,
  Node,
  PluralNode,
  TextNode,
  LiteralNode,
  LinkedNode,
  LinkedKeyNode,
  LinkedModifierNode,
  ListNode,
  NamedNode
} from './nodes'

export const ERROR_DOMAIN = 'minifier'

/* eslint-disable @typescript-eslint/no-explicit-any */

export function minify(node: Node) {
  node.t = node.type

  switch (node.type) {
    case NodeTypes.Resource:
      const resource = node as ResourceNode
      minify(resource.body)
      resource.b = resource.body
      delete (resource as any).body
      break
    case NodeTypes.Plural:
      const plural = node as PluralNode
      const cases = plural.cases
      for (let i = 0; i < cases.length; i++) {
        minify(cases[i])
      }
      plural.c = cases
      delete (plural as any).cases
      break
    case NodeTypes.Message:
      const message = node as MessageNode
      const items = message.items
      for (let i = 0; i < items.length; i++) {
        minify(items[i])
      }
      message.i = items
      delete (message as any).items
      if (message.static) {
        message.s = message.static
        delete (message as any).static
      }
      break
    case NodeTypes.Text:
    case NodeTypes.Literal:
    case NodeTypes.LinkedModifier:
    case NodeTypes.LinkedKey:
      const valueNode = node as
        | TextNode
        | LiteralNode
        | LinkedKeyNode
        | LinkedModifierNode
      if (valueNode.value) {
        valueNode.v = valueNode.value
        delete (valueNode as any).value
      }
      break
    case NodeTypes.Linked:
      const linked = node as LinkedNode
      minify(linked.key)
      linked.k = linked.key
      delete (linked as any).key
      if (linked.modifier) {
        minify(linked.modifier)
        linked.m = linked.modifier
        delete (linked as any).modifier
      }
      break
    case NodeTypes.List:
      const list = node as ListNode
      list.i = list.index
      delete (list as any).index
      break
    case NodeTypes.Named:
      const named = node as NamedNode
      named.k = named.key
      delete (named as any).key
      break
    default:
      if (__DEV__) {
        throw createCompileError(
          CompileErrorCodes.UNHANDLED_MINIFIER_NODE_TYPE,
          null,
          {
            domain: ERROR_DOMAIN,
            args: [node.type]
          }
        )
      }
  }

  delete (node as any).type
}

/* eslint-enable @typescript-eslint/no-explicit-any */
