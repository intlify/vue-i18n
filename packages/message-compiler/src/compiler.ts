import { CompileOptions } from './options'
import { createParser } from './parser'
import { transform } from './transformer'
import { generate, CodeGenResult } from './generator'
import { assign } from '@intlify/shared'

export function baseCompile(
  source: string,
  options: CompileOptions = {}
): CodeGenResult {
  const assignedOptions = assign({}, options)

  // parse source codes
  const parser = createParser(assignedOptions)
  const ast = parser.parse(source)

  // transform ASTs
  transform(ast, assignedOptions)

  // generate javascript codes
  return generate(ast, assignedOptions)
}
