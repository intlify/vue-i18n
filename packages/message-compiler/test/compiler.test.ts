/* eslint-disable @typescript-eslint/no-empty-function, no-irregular-whitespace */

import { baseCompile as compile } from '../src/compiler'

test(`@.caml:{'no apples'} | {0} apple | {n}　apples`, () => {
  const { code, ast, map } = compile(
    `@.caml:{'no apples'} | {0} apple | {n}　apples`
  )
  expect(code).toMatchSnapshot('code')
  expect(ast).toMatchSnapshot('ast')
  expect(map).toMatchSnapshot('map')
})

describe('arrow mode', () => {
  test('hello world', () => {
    const { code } = compile(`hello world`, { mode: 'arrow' })
    new Function(code)
    expect(code).toMatchSnapshot('code')
  })

  test('hello\\n world', () => {
    const { code } = compile(`hello\n world`, { mode: 'arrow' })
    new Function(code)
    expect(code).toMatchSnapshot('code')
  })

  test('hi {0}!', () => {
    const { code } = compile(`hi {0}!`, { mode: 'arrow' })
    new Function(code)
    expect(code).toMatchSnapshot('code')
  })

  test('hi {name}!', () => {
    const { code } = compile(`hi {name}!`, { mode: 'arrow' })
    new Function(code)
    expect(code).toMatchSnapshot('code')
  })

  test("hi { 'kazupon' }!", () => {
    const { code } = compile(`hi { 'kazupon' }!`, { mode: 'arrow' })
    new Function(code)
    expect(code).toMatchSnapshot('code')
  })

  test(`hi @.upper:{'name'} !`, () => {
    const { code } = compile(`hi @.upper:{'name'} !`, { mode: 'arrow' })
    new Function(code)
    expect(code).toMatchSnapshot('code')
  })

  test('!#%^&*()-_+=[]:;?.<>"`', () => {
    const { code } = compile(`hi {'${'!#%^&*()-_+=[]:;?.<>"`'}'} !`, {
      mode: 'arrow'
    })
    new Function(code)
    expect(code).toMatchSnapshot('code')
  })

  test(` | | | `, () => {
    const { code } = compile(` | | | `, { mode: 'arrow' })
    new Function(code)
    expect(code).toMatchSnapshot('code')
  })

  test(`@.caml:{'no apples'} | {0} apple | {n}　apples`, () => {
    const { code } = compile(`@.caml:{'no apples'} | {0} apple | {n}　apples`, {
      mode: 'arrow'
    })
    new Function(code)
    expect(code).toMatchSnapshot('code')
  })
})

describe('edge cases', () => {
  test(` | | | `, () => {
    const { code } = compile(` | | | `, {
      onError(error) {
        expect({ ...error, message: error.message }).toMatchSnapshot('error')
      }
    })
    expect(code).toMatchSnapshot('code')
  })

  test(`hi %s !`, () => {
    const { code } = compile(`hi %s !`)
    expect(code).toMatchSnapshot('code')
  })

  test(`%`, () => {
    const { code } = compile(`%`)
    expect(code).toMatchSnapshot('code')
  })

  test(`no apples %| one apple % |  too much apples  `, () => {
    const { code } = compile(`no apples %| one apple % |  too much apples  `)
    expect(code).toMatchSnapshot('code')
  })

  test(`{_field} with the same value already exists.`, () => {
    const { code } = compile(`{_field} with the same value already exists.`)
    expect(code).toMatchSnapshot('code')
  })

  test(`hi @._upper:{_name} !`, () => {
    const { code } = compile(`hi @._upper:{_name} !`)
    expect(code).toMatchSnapshot('code')
  })
})
