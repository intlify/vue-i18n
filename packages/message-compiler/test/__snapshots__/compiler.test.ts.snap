// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@.caml:{'no apples'} | {0} apple | {n}　apples: ast 1`] = `
Object {
  "body": Object {
    "cases": Array [
      Object {
        "end": 19,
        "items": Array [
          Object {
            "end": 20,
            "key": Object {
              "end": 20,
              "loc": Object {
                "end": Object {
                  "column": 21,
                  "line": 1,
                  "offset": 20,
                },
                "start": Object {
                  "column": 8,
                  "line": 1,
                  "offset": 7,
                },
              },
              "start": 7,
              "type": 9,
              "value": "no apples",
            },
            "loc": Object {
              "end": Object {
                "column": 21,
                "line": 1,
                "offset": 20,
              },
              "start": Object {
                "column": 1,
                "line": 1,
                "offset": 0,
              },
            },
            "modifier": Object {
              "end": 6,
              "loc": Object {
                "end": Object {
                  "column": 7,
                  "line": 1,
                  "offset": 6,
                },
                "start": Object {
                  "column": 2,
                  "line": 1,
                  "offset": 1,
                },
              },
              "start": 1,
              "type": 8,
              "value": "caml",
            },
            "start": 0,
            "type": 6,
          },
        ],
        "loc": Object {
          "end": Object {
            "column": 21,
            "line": 1,
            "offset": 20,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 2,
      },
      Object {
        "end": 26,
        "items": Array [
          Object {
            "end": 26,
            "index": 0,
            "loc": Object {
              "end": Object {
                "column": 27,
                "line": 1,
                "offset": 26,
              },
              "start": Object {
                "column": 24,
                "line": 1,
                "offset": 23,
              },
            },
            "start": 23,
            "type": 5,
          },
          Object {
            "end": 32,
            "loc": Object {
              "end": Object {
                "column": 33,
                "line": 1,
                "offset": 32,
              },
              "start": Object {
                "column": 27,
                "line": 1,
                "offset": 26,
              },
            },
            "start": 26,
            "type": 3,
            "value": " apple",
          },
        ],
        "loc": Object {
          "end": Object {
            "column": 33,
            "line": 1,
            "offset": 32,
          },
          "start": Object {
            "column": 24,
            "line": 1,
            "offset": 23,
          },
        },
        "start": 23,
        "type": 2,
      },
      Object {
        "end": 45,
        "items": Array [
          Object {
            "end": 38,
            "key": "n",
            "loc": Object {
              "end": Object {
                "column": 39,
                "line": 1,
                "offset": 38,
              },
              "start": Object {
                "column": 36,
                "line": 1,
                "offset": 35,
              },
            },
            "start": 35,
            "type": 4,
          },
          Object {
            "end": 45,
            "loc": Object {
              "end": Object {
                "column": 46,
                "line": 1,
                "offset": 45,
              },
              "start": Object {
                "column": 39,
                "line": 1,
                "offset": 38,
              },
            },
            "start": 38,
            "type": 3,
            "value": "　apples",
          },
        ],
        "loc": Object {
          "end": Object {
            "column": 46,
            "line": 1,
            "offset": 45,
          },
          "start": Object {
            "column": 36,
            "line": 1,
            "offset": 35,
          },
        },
        "start": 35,
        "type": 2,
      },
    ],
    "end": 45,
    "loc": Object {
      "end": Object {
        "column": 46,
        "line": 1,
        "offset": 45,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 1,
  },
  "end": 45,
  "helpers": Array [
    "normalize",
    "linked",
    "interpolate",
    "list",
    "named",
    "plural",
  ],
  "loc": Object {
    "end": Object {
      "column": 46,
      "line": 1,
      "offset": 45,
    },
    "source": "@.caml:{'no apples'} | {0} apple | {n}　apples",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`@.caml:{'no apples'} | {0} apple | {n}　apples: code 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize, linked: _linked, interpolate: _interpolate, list: _list, named: _named, plural: _plural } = ctx
  return _plural([
    _normalize([
      _linked(\\"no apples\\", \\"caml\\")
    ]), _normalize([
      _interpolate(_list(0)), \\" apple\\"
    ]), _normalize([
      _interpolate(_named(\\"n\\")), \\"　apples\\"
    ])
  ])
}"
`;

exports[`@.caml:{'no apples'} | {0} apple | {n}　apples: map 1`] = `undefined`;

exports[`arrow mode  | | | : code 1`] = `"(ctx) => {const { normalize: _normalize, plural: _plural } = ctx;return _plural([_normalize([]), _normalize([]), _normalize([]), _normalize([])])}"`;

exports[`arrow mode !#%^&*()-_+=[]:;?.<>"\`: code 1`] = `"(ctx) => {const { normalize: _normalize } = ctx;return _normalize([\\"hi \\", \\"!#%^&*()-_+=[]:;?.<>\\\\\\"\`\\", \\" !\\"])}"`;

exports[`arrow mode @.caml:{'no apples'} | {0} apple | {n}　apples: code 1`] = `"(ctx) => {const { normalize: _normalize, linked: _linked, interpolate: _interpolate, list: _list, named: _named, plural: _plural } = ctx;return _plural([_normalize([_linked(\\"no apples\\", \\"caml\\")]), _normalize([_interpolate(_list(0)), \\" apple\\"]), _normalize([_interpolate(_named(\\"n\\")), \\"　apples\\"])])}"`;

exports[`arrow mode hello world: code 1`] = `"(ctx) => {const { normalize: _normalize } = ctx;return _normalize([\\"hello world\\"])}"`;

exports[`arrow mode hello\\n world: code 1`] = `"(ctx) => {const { normalize: _normalize } = ctx;return _normalize([\\"hello\\\\n world\\"])}"`;

exports[`arrow mode hi @.upper:{'name'} !: code 1`] = `"(ctx) => {const { normalize: _normalize, linked: _linked } = ctx;return _normalize([\\"hi \\", _linked(\\"name\\", \\"upper\\"), \\" !\\"])}"`;

exports[`arrow mode hi { 'kazupon' }!: code 1`] = `"(ctx) => {const { normalize: _normalize } = ctx;return _normalize([\\"hi \\", \\"kazupon\\", \\"!\\"])}"`;

exports[`arrow mode hi {0}!: code 1`] = `"(ctx) => {const { normalize: _normalize, interpolate: _interpolate, list: _list } = ctx;return _normalize([\\"hi \\", _interpolate(_list(0)), \\"!\\"])}"`;

exports[`arrow mode hi {name}!: code 1`] = `"(ctx) => {const { normalize: _normalize, interpolate: _interpolate, named: _named } = ctx;return _normalize([\\"hi \\", _interpolate(_named(\\"name\\")), \\"!\\"])}"`;

exports[`edge cases  | | | : code 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize, plural: _plural } = ctx
  return _plural([
    _normalize([
      
    ]), _normalize([
      
    ]), _normalize([
      
    ]), _normalize([
      
    ])
  ])
}"
`;

exports[`edge cases  | | | : error 1`] = `
Object {
  "code": 10,
  "domain": "parser",
  "location": Object {
    "end": Object {
      "column": 8,
      "line": 1,
      "offset": 7,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "message": "Plural must have messages",
}
`;

exports[`edge cases %: code 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize } = ctx
  return _normalize([
    \\"%\\"
  ])
}"
`;

exports[`edge cases {_field} with the same value already exists.: code 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize, interpolate: _interpolate, named: _named } = ctx
  return _normalize([
    _interpolate(_named(\\"_field\\")), \\" with the same value already exists.\\"
  ])
}"
`;

exports[`edge cases hi %s !: code 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize } = ctx
  return _normalize([
    \\"hi %s !\\"
  ])
}"
`;

exports[`edge cases hi @._upper:{_name} !: code 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize, interpolate: _interpolate, named: _named, linked: _linked } = ctx
  return _normalize([
    \\"hi \\", _linked(_interpolate(_named(\\"_name\\")), \\"_upper\\"), \\" !\\"
  ])
}"
`;

exports[`edge cases no apples %| one apple % |  too much apples  : code 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize, plural: _plural } = ctx
  return _plural([
    _normalize([
      \\"no apples %\\"
    ]), _normalize([
      \\"one apple %\\"
    ]), _normalize([
      \\"too much apples  \\"
    ])
  ])
}"
`;
